<a href="https://colab.research.google.com/github/waj-ha/child-safety-fars-analysis/blob/main/notebooks/260606_import_002.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Analysis Enssential Tools and Libraries:
These libraries are essential tools for most data science workflows:

*   **NumPy**: Fundamental package for numerical computation, especially with arrays and matrices.
*   **Pandas**: Provides high-performance, easy-to-use data structures and data analysis tools (DataFrames).
*   **Matplotlib**: A comprehensive library for creating static, animated, and interactive visualizations.
*   **Seaborn**: Built on Matplotlib, it provides a high-level interface for drawing attractive and informative statistical graphics.
*   **Scikit-learn**: A robust machine learning library featuring various classification, regression, and clustering algorithms, along with tools for model selection and preprocessing.
*   **Statsmodels**: Provides classes and functions for the estimation of many different statistical models, as well as for conducting statistical tests and statistical data exploration.

The `!pip install` command is used to install Python packages. The `!` at the beginning tells Colab to run the command as a shell command. In this example, `numpy` is the name of the package to be installed.

In [118]:
common_data_science_libraries = [
    "numpy",        # Numerical operations
    "pandas",       # Data manipulation and analysis
    "matplotlib",   # Basic plotting
    "seaborn",      # Enhanced statistical data visualization
    "statsmodels",  # Statistical modeling
    "requests",     # HTTP requests for web data
]

# Install all listed libraries
!pip install {" ".join(common_data_science_libraries)}



In [119]:
import numpy as np
import pandas as pd
import matplotlib
import matplotlib.pyplot as plt # Import pyplot for plotting functions
import seaborn as sns
import sklearn
import tensorflow as tf

print(f"NumPy version: {np.__version__}")
print(f"Pandas version: {pd.__version__}")
print(f"Matplotlib version: {matplotlib.__version__}") # Access version from the main module
print(f"Seaborn version: {sns.__version__}")
print(f"Scikit-learn version: {sklearn.__version__}")
print(f"TensorFlow version: {tf.__version__}")

# You can add simple verification tests here if needed, for example:
# print(np.array([1, 2, 3]) + np.array([4, 5, 6]))
# print(pd.DataFrame({'A': [1, 2], 'B': [3, 4]}))

NumPy version: 2.0.2
Pandas version: 2.2.2
Matplotlib version: 3.10.0
Seaborn version: 0.13.2
Scikit-learn version: 1.6.1
TensorFlow version: 2.20.0


### Dataset Import Techniques
Raw data is placed at Github;

https://github.com/waj-ha/child-safety-fars-analysis.git

and accessed, and imported via Pandas Libraries. The files sizes are more than 100MB. So, these are

In [120]:
import pandas as pd
#url_accident = 'https://media.githubusercontent.com/media/waj-ha/child-safety-fars-analysis/refs/heads/main/data/raw/accident.csv'
#df_raw_accident = pd.read_csv(url_accident,encoding='unicode_escape', dtype=str, low_memory=False)
url_person = 'https://media.githubusercontent.com/media/waj-ha/child-safety-fars-analysis/refs/heads/main/data/raw/person.csv'
df_raw_person = pd.read_csv(url_person,encoding='unicode_escape', dtype=str, low_memory=False)
#url_vehicle = 'https://media.githubusercontent.com/media/waj-ha/child-safety-fars-analysis/refs/heads/main/data/raw/vehicle.csv'
#df_raw_vehicle = pd.read_csv(url_vehicle,encoding='unicode_escape', dtype=str, low_memory=False)

### EDA

In [121]:
def dataset_overview(df, dataset_name):
    print("*****************")
    print(f"{dataset_name} Dataset:")
    print("*****************")
    print("\nShape:", df.shape)
    print("\nCount:")
    print(df.nunique().sort_values(ascending=False))
    print("\nData Types:")
    print(df.dtypes.value_counts(ascending=False))
    print("\nDuplicates:", df.duplicated().sum())
    print(f"\nMissing Values Count:")
    missing_values = (df.isnull().sum().sort_values(ascending=False))
    non_zero_missing_values = missing_values[missing_values > 0]
    print(non_zero_missing_values)
    print(f"\nMissing Values Percentage:")
    missing_percentage = round(df.isnull().sum() * 100 / len(df), 2).sort_values(ascending=False)
    non_zero_missing_percentage = missing_percentage[missing_percentage > 0]
    print(non_zero_missing_percentage)
    print("\n---------------------------------------------------\n")

dataset_overview(df_raw_person, "Person")


*****************
Person Dataset:
*****************

Shape: (88326, 126)

Count:
ST_CASE          36297
VPICMODEL         2381
VPICMODELNAME     2325
DEATH_TM          1457
DEATH_TMNAME      1457
                 ...  
EXTRICATNAME         3
FIRE_EXPNAME         2
FIRE_EXP             2
SCH_BUSNAME          2
SCH_BUS              2
Length: 126, dtype: int64

Data Types:
object    126
Name: count, dtype: int64

Duplicates: 0

Missing Values Count:
VPICMAKE             9020
MOD_YEAR             9020
VPICMODEL            9020
VPICMAKENAME         9020
MOD_YEARNAME         9020
FIRE_EXPNAME         9020
FIRE_EXP             9020
IMPACT1NAME          9020
IMPACT1              9020
ROLLOVERNAME         9020
ROLLOVER             9020
EMER_USENAME         9020
EMER_USE             9020
SPEC_USENAME         9020
SPEC_USE             9020
TOW_VEHNAME          9020
TOW_VEH              9020
GVWR_TONAME          9020
GVWR_TO              9020
GVWR_FROMNAME        9020
GVWR_FROM            9020
ICF

Relevant Variables:

**`ST_CASE`**: Identify the crash case and Unique ID.

**`AGE`**: Identify the age groups and risk difference.

**`SEX`**: Compare outcomes in males and females.

**`PER_TYP`**: Identify the type of perssenger. Occupant/Pedestrain.

**`RESTRAINT_USE`**: Compare the impact of use of restaint Systems and seatbelts.

**`SEAT_POS`**: Analyses seat position Effects.

**`INJ_SEV`**: Measure Crash outcomes.

**`BODY_TYP`**: Compare Vehicle types.

**`HOUR`**: Study Time of the day patterns.

**`WEATHER`**: Examine environmental conditions.

**`SPEEDING`**: Assess influence of speed related crashes

**`RURAL_URBAN`**: Compare rural or urban crash characteristics.

**`STATE`**: Compare the US states.


## AGE

From FARS reference manual, we know the outliers are defined with special codes. For **`AGE`** these code are,

*   `998` = Unknow
*   `999` = Not Reported



## Dataset Overview

**The Accident data file** includes crash data.
It contains the data elements **`ST_CASE`** and **`STATE`**,
which are described in the Key Data Elements at the beginning of the Data Element Definitions
and Codes section. The Accident data file also contains the data elements on the following pages.
**`ST_CASE`** is the unique case identifier for each record.


### Dataset Overview

Three main datasets related to traffic accidents:

*   **`df_accident`**: This DataFrame contains accident-level information. It has `36297` rows and `80` columns. Each row likely represents a unique accident, with columns detailing various aspects of the crash.

*   **`df_person`**: This DataFrame contains person-level information involved in the accidents. It has `88326` rows and `126` columns. This dataset provides details about each person involved in the accidents recorded in `df_accident`.

*   **`df_vehicle`**: This DataFrame contains vehicle-level information involved in the accidents. It has `56011` rows and `201` columns. This dataset provides details about each vehicle involved in the accidents recorded in `df_accident`.

### Primary Key Identification

Given the structure of the FARS datasets, the primary keys are likely:

*   For **`df_accident`**: The `ST_CASE` column serves as the primary key, uniquely identifying each accident.

*   For **`df_person`**: A composite primary key consisting of `ST_CASE`, `VEH_NO` (Vehicle Number), and `PER_NO` (Person Number) would uniquely identify each person involved in a specific vehicle in a particular accident.

*   For **`df_vehicle`**: A composite primary key consisting of `ST_CASE` and `VEH_NO` would uniquely identify each vehicle involved in a particular accident.